go/rpcclientlib/obscuro_client.go

```diff
@@ -19,39 +28,264 @@ const (
+		return c.rpcClient.Call(&result, method, args...)
+	}
+
+	// we setup a generic rawResult to receive the response (then we can decrypt it as necessary into the requested result type)
+	var err error
+	var rawResult interface{}
+	if result == nil {
+		// we set result receiver to nil if caller set nil since no return needs to be set
+		rawResult = nil
+	}
+
+	args, err = c.sanitiseArgs(method, args)
+	if err != nil {
+		return err
+	}
+
+	// encode the params into a json blob and encrypt them
+	var encryptedParams []byte
+	encryptedParams, err = c.encryptArgs(args...)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt args for rpc call - %w", err)
+	}
+	err = c.rpcClient.Call(&rawResult, method, encryptedParams)
+	if err != nil {
+		return fmt.Errorf("rpc call failed - %w", err)
+	}
+
+	// if caller not interested in response, we're done
+	if result == nil {
+		return nil
+	}
+
+	// todo: remove this non-decryption part when we make viewing key encryption mandatory across all tests
+	// extract result from the data as-is, in case we can't decrypt/process it below
+	decrypted, ok := rawResult.([]byte) // if viewing key was nil we try and extract result from the data as-is
+	if !ok {
+		decStr, ok := rawResult.(string)
+		if ok {
+			decrypted = []byte(decStr)
+		}
+	}
+	if c.viewingPrivKey != nil {
+		// method is sensitive, so we decrypt it before unmarshalling the result
+		decrypted, err = c.decryptResponse(rawResult)
+		if err != nil {
+			return fmt.Errorf("failed to decrypt args for rpc call - %w", err)
+		}
+	}
+
+	// process the decrypted result to get the desired type and set it on the result pointer
+	err = c.setResult(decrypted, result)
+	if err != nil {
+		return fmt.Errorf("failed to extract result from RPC response: %w", err)
+	}
+
+	return nil
+}
+
+func (c *networkClient) encryptArgs(args ...interface{}) ([]byte, error) {
+	if len(args) == 0 {
+		return nil, nil
+	}
+
+	paramsJSON, err := json.Marshal(args)
+	if err != nil {
+		return nil, fmt.Errorf("could not json encode request params: %w", err)
+	}
+
+	if c.viewingPrivKey == nil {
+		// todo: remove this when viewing keys are non-optional for testing purposes
+		// if no viewing key is set then we don't encrypt (we wouldn't be able to decrypt result).
+		return paramsJSON, nil
 	}
+	return c.encryptParamBytes(paramsJSON)
 }
 
-func (c *clientImpl) Call(result interface{}, method string, args ...interface{}) error {
-	return c.rpcClient.Call(&result, method, args...)
+func (c *networkClient) encryptParamBytes(params []byte) ([]byte, error) {
+	encryptedParams, err := ecies.Encrypt(rand.Reader, c.enclavePublicKey, params, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not encrypt request params with enclave public key: %w", err)
+	}
+	return encryptedParams, nil
+}
+
+func (c *networkClient) decryptResponse(resultBlob interface{}) ([]byte, error) {
+	encryptedResult := common.Hex2Bytes(resultBlob.(string))
+	if c.viewingPrivKey == nil {
```

Comment:
@BedrockSquirrel We've already checked this above, before calling into this method.